Refactor dimensiondata_network module

tintoy opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

module_utils/dimensiondata.py
modules/cloud/dimensiondata/dimensiondata_network.py

##### ANSIBLE VERSION
```
ansible 2.3.0
```

##### SUMMARY

This is a refactoring of ansible/ansible#19325 (and, before that, ansible/ansible-modules-extras#2979).
I apologise for the extra work this will probably create but I believe it is necessary to make this change before any more of the outstanding PRs are brought across.

Please let me know if I've missed anything (I've tried to apply the lessons learned from previous PRs to this one).

tintoy committed:
Refactor dimensiondata_network to use shared base class for common functionality.

tintoy commented:
Hi - I'm not sure I understand what's failing here:

https://app.shippable.com/runs/58966f246e5b460f005d9949/1/console

```
2017-02-05 00:23:31 Sanity check using validate-modules
2017-02-05 00:23:31 Run command: test/sanity/validate-modules/validate-modules lib/ansible/modules/cloud/dimensiondata/dimensiondata_network.py --exclude '^(lib/ansible/modules/utilities/logic/async_status.py|lib/ansible/modules/utilities/helper/_accelerate.py)'
2017-02-05 00:23:32 Traceback (most recent call last):
2017-02-05 00:23:32   File "test/sanity/validate-modules/validate-modules", line 751, in <module>
2017-02-05 00:23:32     main()
2017-02-05 00:23:32   File "test/sanity/validate-modules/validate-modules", line 721, in main
2017-02-05 00:23:32     mv.validate()
2017-02-05 00:23:32   File "test/sanity/validate-modules/validate-modules", line 639, in validate
2017-02-05 00:23:32     self._find_has_import()
2017-02-05 00:23:32   File "test/sanity/validate-modules/validate-modules", line 322, in _find_has_import
2017-02-05 00:23:32     if target.id.lower().startswith('has_'):
2017-02-05 00:23:32 AttributeError: 'Tuple' object has no attribute 'id'
2017-02-05 00:23:32 ERROR: Command "test/sanity/validate-modules/validate-modules lib/ansible/modules/cloud/dimensiondata/dimensiondata_network.py --exclude '^(lib/ansible/modules/utilities/logic/async_status.py|lib/ansible/modules/utilities/helper/_accelerate.py)'" returned exit status 1.
```

Is this failing due to the `HAS_LIBCLOUD` stuff?

tintoy committed:
Experiment: remove the assignments in the "except ImportError:" block that keep PyCharm happy. If this fixes the build, then I reckon there's a bug in the validate-modules script (https://github.com/ansible/ansible/blob/devel/test/sanity/validate-modules/validate-modules#L322).

tintoy committed:
Remove unused imports.

tintoy commented:
Hmm. My import code was wrong, but there may also be a bug in [validate-modules](https://github.com/ansible/ansible/blob/devel/test/sanity/validate-modules/validate-modules#L322) (although admittedly it's a silly edge-case).

Anyway, should be ready for review now :)

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_network.py:
```diff
@@ -159,181 +162,239 @@
+                        type='str', required=False
+                    ),
+                    mcp_password=dict(
+                        type='str', required=False
```
`, no_log=true`

tintoy replied:
Thanks - didn't know about that one!

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_network.py:
```diff
@@ -159,181 +162,239 @@
+        super(DimensionDataNetworkModule, self).__init__(
+            module=AnsibleModule(
+                argument_spec=dict(
+                    region=dict(
```
We have a 160 char limit, so you can put each of the options on a single line, rather than spread over three.

tintoy replied:
I started off like that, but it looked difficult to read (to me, at least). I think it's the use of `dict` instead of `{}`. But it that's the preferred style I'm happy to change it :)

gundalow replied:
Thanks, for changing that, it's now consistent with the other modules.

tintoy commented:
@gundalow I was thinking of appending the common module parameter definitions in the `DimensionDataModule` base class rather than explicitly specifying them for each module. On the plus side, they only need to be declared (and processed) once. But on the minus side, it would mean you can't just see all the parameter definitions in one place. But again on the plus side all parameters are listed in the documentation fields so perhaps it's not really that big a deal.

I think this is my preferred approach, but I'm open leaving the way it is; I figure you've seen enough modules to know what's normal and what's too weird to be easily maintainable by others with common expectations :)

tintoy committed:
Changes based on feedback from @gundalow for ansible/ansible#21043.

- Use no_log=True for mcp_password parameter.
- Collapse module parameter definitions.

gundalow commented:
Regarding common options have a look at module_utils/f5.py and the modules under network/f5 they use `argument_spec = f5_argument_spec()` to pull in common arguments.

You can then document the common args in `docs_fragments`

I've got some PRs in flight that (hopefully) explain this a bit clearer:

https://github.com/ansible/ansible/pull/21021
https://github.com/ansible/ansible/pull/20794

tintoy committed:
Use shared definitions and doc fragments for common module arguments (ansible/ansible#21043).

tintoy commented:
@gundalow - like this? https://github.com/ansible/ansible/pull/21043/commits/81d4a139fb3aa743908bd90f9bc30795ae0b02d8

tintoy commented:
BTW, those updated docs you linked in PRs are super-useful. Thanks! :)

tintoy committed:
Make default network plan "ESSENTIALS", rather than "ADVANCED" (this is consistent with our other tooling).
Tidy up module parameter documentation.

tintoy committed:
Simplify dimensiondata module documentation fragments (didn't know you could include multiple fragments).

gundalow commented:
From a quick scan this looks good. I'll review properly tomorrow. 
Glad my new doc PRs are helpful. Any comments in those are very welcome.

gundalow commented:
That should have said "any review comments on my docs PRs would be very welcome".

gundalow commented on lib/ansible/utils/module_docs_fragments/dimensiondata_with_wait.py:
```diff
@@ -0,0 +1,72 @@
+      - The length of time between successive polls for completion
+    required: false
+    default: 2
+  verify_ssl_cert:
```
To be consistent with other modules can you change this to
`validate_certs`

Remembering to change the code as well.

I wonder if that should go into the "How to document your module" doc

tintoy replied:
Yeah - a glossary of common terms might be useful. I'll fix this up shortly!

gundalow commented on lib/ansible/utils/module_docs_fragments/dimensiondata_with_wait.py:
```diff
@@ -0,0 +1,72 @@
+    default: 2
+  verify_ssl_cert:
+    description:
+      - Check that SSL certificate is valid.
```
If C(no), SSL certificates will not be validated. This should only be used
on personally controlled sites using self-signed certificates.

tintoy replied:
Can `bool` arguments be `yes` / `no`? If not, is `C(false)` acceptable?

tintoy replied:
(have left it as bool for now, but happy to change over if required)

gundalow replied:
type bool should do the "sensible" thing, it uses the following:
`BOOL_TRUE = frozenset([ "true", "t", "y", "1", "yes", "on" ])`

tintoy committed:
Change 'verify_ssl_cert' module parameter to 'validate_certs'.

gundalow commented:
Looks good. FYI, if you raise a PR to update https://github.com/ansible/ansibullbot/blob/master/MAINTAINERS.txt then we can set you (and some other people in your company) as maintainers.

tintoy commented:
Thanks! Will do.

gundalow commented:
Cool, I think we are good to merge. Just give me the nod and I'll do so.

tintoy commented:
Thanks! Yep, should be good to go. I’ll tackle `dimensiondata_vlan` next.
